### ISSUE TITLE
One commit with one commit hash can have multiple commit ids

### ISSUE THREAD

**Issue by bockthom** [se-passau/codeface-extraction#28]
I recently recognized that the number of unique commit ids is greater than the number of unique commit hashes within a project. As this looked weird to me, I tried to find out how this arose.

In fact, in all those cases, the commits which have different ids are completely identical. That is, identical hash, identical author, identical date, etc. So, one commit was just assigned to two different commit ids. As those identical commits with different ids appear roughly every three month, I have a guess where this comes from: As Codeface analyzes three month windows, the commit at the beginning (or end) of a three month range belongs to two subsequent ranges and is therefore added to the database twice -- once for the end of the preceding range and once for the beginning of the subsequent range, ending up in two entries with different ids for the same commit.

As far as we don't use the `id` column of the commit data and only rely on the hashes, this is not a problem. However, when someone uses the `id` instead of the hash, then some commits are used twice. We definitively should fix that, the only question is: Where and how?

Here are several options:
(1) Don't extract the `id` column of commits any more. Then we don't have this problem any more. However, this leads to a breaking change in coronet, as reading `commit.list` files will change.
(2) Still extract the `id` column of commits, but check for duplicated commits and keep only one of them.
(3) Don't change the extraction, but adjust the tools which use the data and filter the duplicated commits there. However, this could be dangerous if the end users are not aware of that problem.

I personally would go for option (1) or (2). Any suggestions @clhunsen?

**Comment by clhunsen** [comment]
Ah, that is a really interesting finding. I was quite sure that the selection of commits for a release range is exclusive at the end. :sweat_smile: Thank you for digging into this. :+1: 

I guess the best solution is to remove the duplicated data as soon as possible. As we do not want to run Codeface analyses again or break coronet regarding to existing data, I would like to see Option (2) implemented. 

-------------------------

I had some thoughts about an implementation [here](https://github.com/se-passau/codeface-extraction/blob/8841ce50c9d2a7f99bb1b8d6aa0973c9f4c70f1e/codeface_extraction/extractions.py#L297-L306):
1. Obtain a tuple `(commit.id, commit.hash)` from each line in the commit data and, then, remove duplicates.
2. Group the remaining tuples by their hashes (see [here](https://stackoverflow.com/a/6602441)).
3. For each group of tuples with a length > 1, get all but the first commit ID.
4. Remove all lines in the original data that correspond to the identified commit ID from Step 3.

Furthermore, I just noticed that we may want to think about a proper inheritance tree, e.g., `CommitExtraction` could be a superclass of `CommitRangeExtraction`. Just a thought.

**Comment by bockthom** [comment]
> I would like to see Option (2) implemented.

Yeah, that sounds like a good idea. I will try to work on that next week (if I find enough time for that).

**Comment by bockthom** [comment]
I have worked on that and found an even simpler solution:

We can solve the problem already when extracting the data from the database, in the SQL statement. We just need to adjust two things:

We group the selected rows by the information which is unique for each line, that is, every line in the outcome can be identified by the combination of commit hash, file, entity id, and entity type. So, when we group the data by those columns, we can aggregate the data for each group (which represents one line in the resulting data). When we take the minimum of the commit ids of such a group, we only keep one of those for each group - and that is exactly what we wanted to have.

While implementing and testing that, I also found out that not only the commits that lay directly on the border between to Codeface ranges (that is, described the start or end of such a range) are twice in the database, but also additional commits which have the same commit date as those. That's just an interesting finding -- such duplications are also removed within the above described procedure.
